api/v1alpha2/azurecluster_conversion.go

```diff
@@ -0,0 +1,122 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha2
+
+import (
+	apiconversion "k8s.io/apimachinery/pkg/conversion"
+	infrav1alpha3 "sigs.k8s.io/cluster-api-provider-azure/api/v1alpha3"
+	utilconversion "sigs.k8s.io/cluster-api/util/conversion"
+	"sigs.k8s.io/controller-runtime/pkg/conversion"
+)
+
+// ConvertTo converts this AzureCluster to the Hub version (v1alpha3).
+func (src *AzureCluster) ConvertTo(dstRaw conversion.Hub) error { // nolint
+	dst := dstRaw.(*infrav1alpha3.AzureCluster)
+
+	if err := Convert_v1alpha2_AzureCluster_To_v1alpha3_AzureCluster(src, dst, nil); err != nil {
+		return err
+	}
+
+	// Manually convert Status.APIEndpoints to Spec.ControlPlaneEndpoint.
+	if len(src.Status.APIEndpoints) > 0 {
```

Comment:
@vincepri @ncdc The roundtripping discussion we had earlier this week is making me wonder if we are potentially doing a lossy conversion anywhere we are handling v1alpha2 -> v1lapha3 conversion involving APIEndpoints like here. If we get a v1alpha2 data type with multiple APIEndpoints, we are basically throwing away the additional data as we convert it to the storage version...  Do we need to potentially persist that additional data via annotation until we drop support for v1alpha2?

I suspect this is something that would pop up if we were handling v1alpha2 -> v1alpha3 -> v1alpha2 roundtripping more automatically in the fuzzing testing.

Reply:
I'd recommend reading https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api_changes.md thoroughly. We may need to make some changes before we ship. For example:

> Do not add a new API version of an existing resource and make it the preferred version in the same release, and do not make it the storage version. The latter is necessary so that a rollback of the apiserver doesn't render resources in etcd undecodable after rollback.

I'm still going through it for a refresher.